Making the HtmlHelper.GetEnumSelectList take GroupName into consideration when creating the list.

stebet opened the pull request:
aspnet/Mvc#2292

Adding the DisplayAttribute.GroupName into the mix when using HtmlHelper.GetEnumSelectList for grouping of items in the list.

No sorting by default.

dnfclas commented:
Hi __@stebet__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, DNFBOT;

dougbu commented on src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelMetadata.cs:
```diff
@@ -137,14 +137,15 @@ public string PropertyName
+        /// The second item in the tuple is the display name.
         /// </value>
-        public abstract IEnumerable<KeyValuePair<string, string>> EnumDisplayNamesAndValues { get; }
+        public abstract IEnumerable<KeyValuePair<Tuple<string, string>, string>> EnumGroupedDisplayNamesAndValues { get; }
```
Realize it's yet-another abstraction but would prefer something more obvious here, not `Tuple`. Suggest something like `public struct EnumGroupAndName { }` and make it immutable.

dougbu commented on src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs:
```diff
@@ -1083,14 +1083,25 @@ protected virtual IEnumerable<SelectListItem> GetEnumSelectList([NotNull] ModelM
                     Value = keyValuePair.Value,
                 };
 
+                if(string.IsNullOrEmpty(keyValuePair.Key.Item1) == false)
```
- Match the style of other code in this repo: Space after `if`.
- Don't compare a non-`Nullable` value with `false`. Just `!string.IsNullOrEmpty(keyValuePair.Key.Item1)`.

dougbu commented on src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelMetadata.cs:
```diff
@@ -137,14 +137,15 @@ public string PropertyName
-        /// An <see cref="IEnumerable{KeyValuePair{string, string}}"/> of mappings between <see cref="Enum"/> field names
-        /// and values. <c>null</c> if <see cref="IsEnum"/> is <c>false</c>.
+        /// An <see cref="IEnumerable{KeyValuePair{Tuple{string, string}, string}"/> of mappings between <see cref="Enum"/> field groups, names
+        /// and values. <c>null</c> if <see cref="IsEnum"/> is <c>false</c>. The first item in the tuple is the group name (empty string if not defined).
```
Less important but some of these are pretty long, making it harder to review. GitHub runs out of room at around 120 characters. (Even less important in test code because tests tend to contain a lot of boiler-plate.)

dougbu commented:
:watch:

stebet commented:
Thanks for the comments.I'll make these changes and update the request.

dougbu commented on src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -0,0 +1,32 @@
+        /// </summary>
+        /// <param name="group">The group name.</param>
+        /// <param name="name">The name.</param>
+        public EnumGroupAndName(string group, string name)
```
- Move constructor to the top.
- Should throw an `ArgumentNullException` if either `group` or `name` are `null`. Code currently calling this is written defensively but using code is not.

dougbu commented on src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -0,0 +1,32 @@
+namespace Microsoft.AspNet.Mvc.ModelBinding
+{
+    /// <summary>
+    /// An abstraction used when grouping enum values for <see cref="ModelMetadata.EnumGroupedDisplayNamesAndValues"/>
```
Minor (I'll do it after taking this): All doc comments must end with a period.

dougbu commented on src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs:
```diff
@@ -1083,14 +1083,25 @@ protected virtual IEnumerable<SelectListItem> GetEnumSelectList([NotNull] ModelM
 
+                if (!string.IsNullOrEmpty(keyValuePair.Key.Group))
+                {
+                    if(!groupList.ContainsKey(keyValuePair.Key.Group))
```
Minor (I'll clean up): Space after `if`.

dougbu commented:
:watch: Very close.

Also:
- Please rebase your fork as well as address last couple of non-minor comments.
- Use an "updated" comment to let me know when you're done. GitHub doesn't notify anyone about new commits added to a PR. (Still sorry I missed fact you addressed all of my comments two days ago.)

stebet commented:
I´ll fix these in a jiffy :)

thanks.

stebet commented:
updated

stebet commented:
Hmm.. did I rebase correctly? I added aspnet/as upstream, fetched latest upstream and rebased from there.

Wondering because the pull request seems huge now :/

dougbu commented:
Looks like you merged `aspnet/dev` rather than fetching the latest then `git rebase` of your fork. After a force push (`git push -f origin`), this PR should show only your commits.

stebet commented:
```git push -f origin``` says everything is up to date :/

stebet commented:
Ok, think I managed to revert back to before the rebase/merge. Now to do this properly.

stebet commented:
updated

I hope this makes sense now. Need to master my git-fu a bit more :)

NeelBhatt commented:
Just curious to know that, Can not we use "Easy property initialize-rs" of C# 6.0 here https://github.com/stebet/Mvc/commit/bb6620bf03feb212da139c70d8fbaf9a1afd4d5a?

dougbu commented:
13af6f8
